extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduDataIndexRead.java

```diff
@@ -0,0 +1,108 @@
+
+import org.locationtech.geowave.core.store.CloseableIterator;
+import org.locationtech.geowave.core.store.CloseableIteratorWrapper;
+import org.locationtech.geowave.core.store.entities.GeoWaveRow;
+import org.locationtech.geowave.datastore.kudu.KuduDataIndexRow;
+import org.locationtech.geowave.datastore.kudu.KuduDataIndexRow.KuduDataIndexField;
+import org.locationtech.geowave.datastore.kudu.util.KuduUtils;
+import org.apache.kudu.Schema;
+import org.apache.kudu.client.KuduScanner;
+import org.apache.kudu.client.KuduPredicate;
+import org.apache.kudu.client.KuduPredicate.ComparisonOp;
+import org.apache.kudu.client.KuduTable;
+import org.apache.kudu.client.RowResult;
+import org.apache.kudu.client.KuduScanner.KuduScannerBuilder;
+import org.apache.kudu.shaded.com.google.common.collect.Lists;
+import org.apache.kudu.client.RowResultIterator;
+import org.locationtech.geowave.core.index.ByteArray;
+import com.google.common.base.Function;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Streams;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+public class KuduDataIndexRead<T> {
+  private final Schema schema;
+  private final short adapterId;
+  final byte[][] dataIds;
+  private final KuduTable table;
+  private final KuduOperations operations;
+  private final boolean visibilityEnabled;
+  private final Predicate<GeoWaveRow> filter;
+  private List<RowResultIterator> results;
+
+  protected KuduDataIndexRead(
+      final short adapterId,
+      final byte[][] dataIds,
+      final KuduTable table,
+      final KuduOperations operations,
+      final boolean visibilityEnabled,
+      final Predicate<GeoWaveRow> filter) {
+    this.adapterId = adapterId;
+    this.dataIds = dataIds;
+    this.table = table;
+    this.schema = table.getSchema();
+    this.operations = operations;
+    this.visibilityEnabled = visibilityEnabled;
+    this.filter = filter;
+    this.results = new ArrayList<>();
+  }
+
+  public CloseableIterator<T> results() {
+    results = new ArrayList<>();
+    KuduPredicate adapterIdPred =
+        KuduPredicate.newComparisonPredicate(
+            schema.getColumn(KuduDataIndexField.GW_ADAPTER_ID_KEY.getFieldName()),
+            ComparisonOp.EQUAL,
+            adapterId);
+    KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
+    KuduScanner scanner;
+    if (dataIds != null) {
+      KuduPredicate partitionPred =
+          KuduPredicate.newInListPredicate(
+              schema.getColumn(KuduDataIndexField.GW_PARTITION_ID_KEY.getFieldName()),
+              Lists.newArrayList(dataIds));
+      scanner = scannerBuilder.addPredicate(partitionPred).addPredicate(adapterIdPred).build();
+    } else {
+      scanner = scannerBuilder.addPredicate(adapterIdPred).build();
```

Comment:
Since both cases use the `adapterIdPred`, could you refactor this to add the `adapterIdPred` before the `if` block and then build the scanner after the `if` block? Then, we can get rid of the `else` block.